### ISSUE TITLE
A suggestion to fix the missing character bug which appears when chatting with RWKV model using Chinese 

### ISSUE THREAD

**Issue by GralchemOz** [oobabooga/text-generation-webui#2447]
**Description**

When chatting with the RWKV model using tgwebui, there's a bug  that causes the output to miss characters if it contains Chinese  characters. This is due to the tokenization of Chinese characters.  Each character always needs two tokens. Therefore, when stream output is  enabled, the first token of a character could be decoded into a  meaningless garbled character and won't become a complete character  until the next token is ready.  
I saw that the code of the modules/RWKV.py file simply discards the detected '\ufffd' garbled characters (in line 120 of RWKV.py). This is the source of this bug because if this is discarded, the corresponding Chinese character is also discarded. 
The method to fix this is very simple. When a '\ufffd' code is detected, just stop the output and decode two tokens at the same time in the next output cycle. 

**Additional Context**

Here is an example that revise the `generate_from_cached_state` method in the RWKV.py files which references the ChatRWKV code:

    def generate_from_cached_state(self, ctx="", token_count=20, args=None, callback=None):
        all_tokens = []
        out_str = ''
        occurrence = {}
        state = copy.deepcopy(self.cached_model_state) if self.cached_model_state is not None else None
        # if we ended up with an empty context, just reuse the cached logits
        # this can happen if a user undoes a message and then sends the exact message again
        # in that case the full context ends up being the same as the cached_context, so the remaining context is empty.
        if ctx == "":
            out = self.cached_output_logits

        for i in range(token_count):
            # forward
            tokens = self.pipeline.encode(ctx) if i == 0 else [token]
            while len(tokens) > 0:
                out, state = self.model.forward(tokens[:args.chunk_len], state)
                tokens = tokens[args.chunk_len:]

            if i == 0:
                begin_token= len(all_tokens)
                last_token_posi=begin_token

            # cache the model state after scanning the context
            # we don't cache the state after processing our own generated tokens because
            # the output string might be post-processed arbitrarily. Therefore, what's fed into the model
            # on the next round of chat might be slightly different what what it output on the previous round
            if i == 0:
                self.cached_context += ctx
                self.cached_model_state = copy.deepcopy(state)
                self.cached_output_logits = copy.deepcopy(out)

            # adjust probabilities
            for n in args.token_ban:
                out[n] = -float('inf')

            for n in occurrence:
                out[n] -= (args.alpha_presence + occurrence[n] * args.alpha_frequency)

            # sampler
            token = self.pipeline.sample_logits(out, temperature=args.temperature, top_p=args.top_p, top_k=args.top_k)
            if token in args.token_stop:
                break

            all_tokens += [token]
            if token not in occurrence:
                occurrence[token] = 1
            else:
                occurrence[token] += 1

            # output
            tmp = self.pipeline.decode(all_tokens[last_token_posi:])
            if '\ufffd' not in tmp:  # is valid utf-8 string?
                if callback:
                    callback(tmp)
                    
                out_str += tmp
                last_token_posi = begin_token + i + 1  
        return out_str

This is the first time I have raised an issue, so please let me know if anything is incorrect.

**Comment by Starlento** [comment]
Your fix helped me. I really appreciate that and hope this could be merge by a pull request.

**Comment by matatonic** [comment]
I had the same bug when I was coding the openai extension when the output contained a lot of emoji.
I handled it a bit differently - I just continued, letting it buffer, and don't output that value. It's a partial unicode character but I think it's not really in the data, but when you ask for a character it has to return something, so for my case it was safe to just ignore it and continue buffering.

```
                if req_params['stream']:
                    # Streaming
                    new_content = answer[len_seen:]

                    if not new_content or chr(0xfffd) in new_content:  # partial unicode character, don't send it yet.
                        continue

                    seen_content = answer
```

https://github.com/oobabooga/text-generation-webui/blob/main/extensions/openai/script.py#LL476C43-L476C43

**Comment by oobabooga** [comment]
Brilliant @matatonic, I have added this change here https://github.com/oobabooga/text-generation-webui/commit/d183c7d29ecd37cb12411aad50d089933dc7387d

It also fixes https://github.com/oobabooga/text-generation-webui/issues/1603

**Comment by GralchemOz** [comment]
> Brilliant @matatonic, I have added this change here [d183c7d](https://github.com/oobabooga/text-generation-webui/commit/d183c7d29ecd37cb12411aad50d089933dc7387d)
> 
> It also fixes #1603
This bug still remains according to our test. Due to the special nature of RWKV model, only change the RWKV,py could fix it .......

**Comment by GralchemOz** [comment]
@oobabooga This bug still remains according to our test. Due to the special nature of RWKV model, only change the RWKV,py could fix it .......

**Comment by coyude** [comment]
> @oobabooga This bug still remains according to our test. Due to the special nature of RWKV model, only change the RWKV,py could fix it .......
Yes, the bug has not been fixed. Only by modifying the RWKV.py file under the "modules" folder can it be resolved. 🤔@oobabooga

**Comment by coyude** [comment]
> @oobabooga This bug still remains according to our test. Due to the special nature of RWKV model, only change the RWKV,py could fix it .......

Yes, the bug has not been fixed. Only by modifying the RWKV.py file under the "modules" folder can it be resolved. 🤔
